_posts/2024-04-26-modules-101.markdown

```diff
@@ -0,0 +1,135 @@
+
+Modify `Cargo.toml` to specify crate-type.  Read [Rust docs](https://doc.rust-lang.org/reference/linkage.html) to understand more about `crate-type`.
+
+```
+[lib]
+crate-type = ["cdylib"]
+```
+
+Now in `src/lib.rs` replace the existing code with the following:
+
+```rust
+#[macro_use]
+extern crate valkey_module;
+
+use valkey_module::{Context, ValkeyResult, ValkeyString, ValkeyValue};
+
+fn hello(_ctx: &Context, _args: Vec<ValkeyString>) -> ValkeyResult {
+    Ok(ValkeyValue::SimpleStringStatic("world2"))
+}
+
+valkey_module! {
+    name: "module2",
+    version: 1,
+    allocator: (valkey_module::alloc::ValkeyAlloc, valkey_module::alloc::ValkeyAlloc),
+    data_types: [],
+    commands: [
+        ["module2.hello", hello, "", 0, 0, 0],
+    ]
+}
+```
+
+Rust syntax is a bit different than C but we are creating `module2` with command `hello` that returns `world2` string.  We are using external crate `valkey_module` with [Rust macros](https://doc.rust-lang.org/book/ch19-06-macros.html) and passing it variables like `name` and `version`.  Some variables like `data_types` and `commands` are arrays and we can pass zero, one or many values.  Since we are not using ctx or args we prefix them with `_` (Rust convention) instead of using `VALKEYMODULE_NOT_USED` as we did in C.  
+
+Run `cargo build` in the root folder.  We will now see `target/debug/libmodule2.dylib` (on macOS).  Build will produce *.so files on Linux and *.dll files on Windows.  
+
+
+## Run Valkey server with both modules
+
+Go back into the Valkey repo folder and run make to compile the Valkey code.  Then add these lines to the bottom of the `valkey.conf` file.
+
+```
+loadmodule path_to_valkey/src/modules/module1.so
+loadmodule path_to_module2/target/debug/libmodule2.dylib
+```
+
+and run `src/valkey-server valkey.conf`.  You will see these messages in the log output.  
+
+```
+Module 'module1' loaded from path_to_valkey/src/modules/module1.so
+...
+Module 'module2' loaded from path_to_module2/target/debug/libmodule2.dylib
+```
+
+Then use `src/valkey-cli` to connect.  
+
+```bash
+src/valkey-cli -3
+127.0.0.1:6379> module list
+1) 1# "name" => "module2"
+   2# "ver" => (integer) 1
+   3# "path" => "path_to_module2/target/debug/libmodule2.dylib"
+   4# "args" => (empty array)
+2) 1# "name" => "module1"
+   2# "ver" => (integer) 1
+   3# "path" => "path_to_valkey/src/modules/module1.so"
+   4# "args" => (empty array)
+127.0.0.1:6379> module1.hello
+world1
+127.0.0.1:6379> module2.hello
+world2
+```
+
+We can now run both modules side by side and if we modify either C or RS file, recompile the code and restart `valkey-server` we will get the new functionality.  Alternatively to specifying modules in .conf file we can use MODULE LOAD and UNLOAD from valkey-cli at runtime.  
```

Comment:
```suggestion
We can now run both modules side by side and if we modify either C or RS file, recompile the code and restart `valkey-server` we will get the new functionality.  Alternatively to specifying modules in .conf file we can use MODULE LOAD and UNLOAD from valkey-cli at runtime.  
```

I don't like blogs that say, "Alternatively do X/Y" without explaining it. So would prefer to either drop the reference or show an example of it.